Building with Java 21 fails due to ancient Antlr plugin

I know, that support for Java 17 is planned for version 7.0, see the [blog post "January 2023 Releases" from 2023/01/15](https://exist-db.org/exist/apps/wiki/blogs/eXist/january-2023-releases) - over a year ago. 
Since Java 8 is EOL for non paying customers for [about two years now](https://www.oracle.com/java/technologies/java-se-support-roadmap.html) and the current LTS is already Java 21, the following should be fixed in the development branch ASAP:

```
[ERROR] Failed to execute goal org.codehaus.mojo:antlr-maven-plugin:2.2:generate (default) on project exist-core: Execution default of goal org.codehaus.mojo:antlr-maven-plugin:2.2:generate failed: The Security Manager is deprecated and will be removed in a future release -> [Help 1]
```

Enviroment:
```
mvn --version
Apache Maven 3.9.6 (bc0240f3c744dd6b6ec2920b3cd08dcc295161ae)
Maven home: /opt/homebrew/Cellar/maven/3.9.6/libexec
Java version: 21.0.1, vendor: Oracle Corporation, runtime: /Users/cmahnke/.sdkman/candidates/java/21.0.1-graal
Default locale: de_DE, platform encoding: UTF-8
OS name: "mac os x", version: "14.2.1", arch: "aarch64", family: "mac"
```

The problem is  introduced by the Maven plugin `org.codehaus.mojo:antlr-maven-plugin:2.2` (which was released [2010](https://mvnrepository.com/artifact/org.codehaus.mojo/antlr-maven-plugin/2.2)) in the `exist-core` module. Even though there seems to be a [fix](https://bugs.launchpad.net/ubuntu/+source/antlr-maven-plugin/+bug/2052993). A clean solution would get rid of this unmaintained dependency of the build system and maybe switch to AntLR 4 in the process.

--- a/.github/workflows/ci-test.yml
+++ b/.github/workflows/ci-test.yml
@@ -39,6 +39,7 @@ jobs:
       fail-fast: false
       matrix:
         os: [ubuntu-latest, windows-latest, macOS-latest]
+        jvm: ['17', '21']
     runs-on: ${{ matrix.os }}
     steps:
       - uses: actions/checkout@v4
@@ -46,7 +47,7 @@ jobs:
         uses: actions/setup-java@v4
         with:
           distribution: temurin
-          java-version: ${{ env.DEV_JDK }}
+          java-version: ${{ matrix.jvm }}
           cache: 'maven'
       - name: Install Maven Daemon
         id: install-mvnd

--- a/.github/workflows/ci-test.yml
+++ b/.github/workflows/ci-test.yml
@@ -78,7 +78,7 @@ jobs:
         if: always()
         uses: actions/upload-artifact@v4
         with:
-          name: ${{ runner.os }}-build-logs
+          name: ${{ runner.os }}-${{ matrix.jvm }}-build-logs
           retention-days: 5
           path: |
             **/*.jfr

--- a/exist-core/pom.xml
+++ b/exist-core/pom.xml
@@ -1067,9 +1067,9 @@ The BaseX Team. The original license statement is also included below.]]></pream
             </plugin>
 
             <plugin>
-                <groupId>org.codehaus.mojo</groupId>
+                <groupId>org.omnifaces</groupId>
                 <artifactId>antlr-maven-plugin</artifactId>
-                <version>2.2</version>
+                <version>2.3</version>
                 <configuration>
                     <grammars>
                         org/exist/xquery/parser/XQuery.g,org/exist/xquery/parser/XQueryTree.g,org/exist/xquery/parser/DeclScanner.g,org/exist/xquery/xqdoc/parser/XQDocParser.g

--- a/extensions/indexes/range/pom.xml
+++ b/extensions/indexes/range/pom.xml
@@ -24,7 +24,7 @@
 -->
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
     <modelVersion>4.0.0</modelVersion>
-    
+
     <parent>
         <groupId>org.exist-db</groupId>
         <artifactId>exist-parent</artifactId>
@@ -42,8 +42,8 @@
         <connection>scm:git:https://github.com/exist-db/exist.git</connection>
         <developerConnection>scm:git:https://github.com/exist-db/exist.git</developerConnection>
         <url>scm:git:https://github.com/exist-db/exist.git</url>
-      <tag>HEAD</tag>
-  </scm>
+        <tag>HEAD</tag>
+    </scm>
 
     <dependencies>
         <dependency>
@@ -67,7 +67,7 @@
             <artifactId>lucene-core</artifactId>
             <version>${lucene.version}</version>
         </dependency>
-        
+
         <dependency>
             <groupId>org.apache.lucene</groupId>
             <artifactId>lucene-analyzers-common</artifactId>
@@ -113,6 +113,13 @@
             <scope>test</scope>
         </dependency>
 
+        <dependency>
+            <groupId>io.github.hakky54</groupId>
+            <artifactId>logcaptor</artifactId>
+            <version>2.9.3</version>
+            <scope>test</scope>
+        </dependency>
+
     </dependencies>
 
     <build>
@@ -128,6 +135,26 @@
         </testResources>
 
         <plugins>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-surefire-plugin</artifactId>
+                <configuration>
+                    <classpathDependencyExcludes>
+                        <classpathDependencyExclude>org.apache.logging.log4j:log4j-slf4j-impl</classpathDependencyExclude>
+                        <classpathDependencyExclude>org.apache.logging.log4j:log4j-slf4j2-impl</classpathDependencyExclude>
+                    </classpathDependencyExcludes>
+                </configuration>
+            </plugin>
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-failsafe-plugin</artifactId>
+                <configuration>
+                    <classpathDependencyExcludes>
+                        <classpathDependencyExclude>org.apache.logging.log4j:log4j-slf4j-impl</classpathDependencyExclude>
+                        <classpathDependencyExclude>org.apache.logging.log4j:log4j-slf4j2-impl</classpathDependencyExclude>
+                    </classpathDependencyExcludes>
+                </configuration>
+            </plugin>
             <plugin>
                 <groupId>org.apache.maven.plugins</groupId>
                 <artifactId>maven-dependency-plugin</artifactId>

--- a/extensions/indexes/range/src/test/java/org/exist/indexing/range/RangeIndexConfigTest.java
+++ b/extensions/indexes/range/src/test/java/org/exist/indexing/range/RangeIndexConfigTest.java
@@ -21,25 +21,23 @@
  */
 package org.exist.indexing.range;
 
+import nl.altindag.log.LogCaptor;
 import org.apache.logging.log4j.Logger;
 import org.easymock.Capture;
-import org.exist.util.JDKCompatibility;
 import org.junit.Test;
 import org.w3c.dom.Document;
 import org.w3c.dom.Element;
 import org.w3c.dom.Node;
 import org.w3c.dom.NodeList;
 
-import java.lang.reflect.Field;
-import java.lang.reflect.Modifier;
 import java.util.HashMap;
 import java.util.Map;
 
-import static junit.framework.TestCase.assertTrue;
 import static org.easymock.EasyMock.*;
 import static org.exist.collections.CollectionConfiguration.DEFAULT_COLLECTION_CONFIG_FILE;
 import static org.exist.indexing.lucene.LuceneIndexConfig.MATCH_ATTR;
 import static org.exist.indexing.lucene.LuceneIndexConfig.QNAME_ATTR;
+import static org.junit.Assert.assertTrue;
 
 public class RangeIndexConfigTest {
 
@@ -88,30 +86,20 @@ public class RangeIndexConfigTest {
         replay(mockConfigNodes, mockConfigNode, mockCreates, mockCreateDocument, mockCreate, mockEmptyNodeList, mockLogger);
 
 
-
         final Map<String, String> namespaces = new HashMap<>();
         namespaces.put("tei", "http://www.tei-c.org/ns/1.0");
 
-        overrideLogger(RangeIndexConfig.class, mockLogger);
+        LogCaptor logCaptor = LogCaptor.forClass(RangeIndexConfig.class);
+
         final RangeIndexConfig config = new RangeIndexConfig(mockConfigNodes, namespaces);
 
-        assertTrue(errorMsgCapture.getValue().contains("Illegal QName: '" + badCreateQName + "'.. QName is invalid: INVALID_LOCAL_PART"));
-        assertTrue(errorMsgCapture.getValue().contains("(" + mockCollectionXConfUri + ")"));
+        assertTrue(logCaptor.getLogs().get(0)
+                .contains("Illegal QName: '" + badCreateQName + "'.. QName is invalid: INVALID_LOCAL_PART"));
 
-        verify(mockConfigNodes, mockConfigNode, mockCreates, mockCreateDocument, mockCreate, mockEmptyNodeList, mockLogger);
+        assertTrue(logCaptor.getLogs().get(0)
+                .contains("(" + mockCollectionXConfUri + ")"));
+
+        verify(mockConfigNodes, mockConfigNode, mockCreates, mockCreateDocument, mockCreate, mockEmptyNodeList);
     }
 
-    private void overrideLogger(final Class clazz, final Logger logger) throws NoSuchFieldException, IllegalAccessException {
-        final Field loggerField = clazz.getDeclaredField("LOG");
-
-        // allow access to private field
-        loggerField.setAccessible(true);
-
-        // remove final modifier
-        final Field modifiersField = JDKCompatibility.getModifiersField();
-        modifiersField.setAccessible(true);
-        modifiersField.setInt(loggerField, loggerField.getModifiers() & ~Modifier.FINAL);
-
-        loggerField.set(null, logger);
-    }
 }

--- a/extensions/modules/simpleql/pom.xml
+++ b/extensions/modules/simpleql/pom.xml
@@ -67,9 +67,9 @@
     <build>
         <plugins>
             <plugin>
-                <groupId>org.codehaus.mojo</groupId>
+                <groupId>org.omnifaces</groupId>
                 <artifactId>antlr-maven-plugin</artifactId>
-                <version>2.2</version>
+                <version>2.3</version>
                 <configuration>
                     <grammars>org/exist/xquery/modules/simpleql/SimpleQLParser.g</grammars>
                 </configuration>

